docs: Cluster Peering for Kubernetes Technical Preview (#13448)
hashicorp/consul: website/content/docs/connect/cluster-peering/k8s.mdx

---
layout: docs
page_title: Cluster Peering on Kubernetes
description: >-
  This page describes how to create peering connections, deploy services, export cluster services, and end peering connections for Consul cluster peering using Kubernetes (K8s).
---

# Cluster Peering on Kubernetes

~> This page covers features that are currently in _technical preview_. Features and functionality are subject to change. You should never use the technical preview release in secure environments or production scenarios. Features in technical preview may face performance and scaling issues, with limited support options available.

To establish a cluster peering connection on Kubernetes, you need to enable the feature in the Helm chart and create custom resource definitions for each side of the peering.

The following Custom Resource Definitions (CRDs) are used to create and manage a peering connection:

- `PeeringAcceptor`: Generates a peering token and accepts an incoming peering connection.
- `PeeringDialer`: Uses a peering token to make an outbound peering connection with the cluster that generated the token.

## Prerequisites

> To create and use cluster peering connections with Kubernetes, you need at least two Kubernetes clusters running in a flat network with Consul on Kubernetes v. 0.45 or later.

Reviewer: ```suggestion
To create and use cluster peering connections with Kubernetes, you need at least two Kubernetes clusters running in a flat network with Consul on Kubernetes v0.0.45 or later.
```

### Helm chart configuration
To establish cluster peering through Kubernetes, deploy clusters with the following Helm values.

```yaml
global:
  peering:
    enabled: true
connectInject:
  enabled: true
meshGateway:
  enabled: true
  replicas: 1
```

## Create a peering connection

To peer Kubernetes clusters running Consul, you need to create a peering token and share it with the other cluster.

> 

Reviewer 2: ```suggestion

### Cluster 1
```

Reviewer 2: This is a suggestion to add a subheader per cluster to be extra clear, but feel free to not use if you think it'll clutter the docs

1. In “cluster-01,” create the `PeeringAcceptor` custom resource.

    <CodeBlockConfig filename="acceptor.yml">

    ```yaml
    apiVersion: consul.hashicorp.com/v1alpha1
    kind: PeeringAcceptor
    metadata:
      name: cluster-02 ## The name of the peer you want to connect to
    spec:
      peer:
        secret:
          name: "peering-token"
          key: "data"
          backend: "kubernetes"
    ```

    </CodeBlockConfig>

1. Apply the `PeeringAcceptor` resource to the first cluster.

    ```shell-session
    $ kubectl apply -f acceptor.yml
    ````

1. Save your peering token so that you can export it to the other cluster.

    ```shell-session

>     $ kubectl get secret peering-token -o yaml > peering-token.yml

Reviewer 3: ```suggestion
    $ kubectl get secret peering-token --output yaml > peering-token.yml
```

    ```

> 

Reviewer 2: ```suggestion

### Cluster 2
```

Reviewer 2: Similarly, feel free to not use if you think it'll clutter the docs

1. Apply the peering token to the second cluster.

    ```shell-session

>     $ kubectl apply -f peering-token.yml

Reviewer 3: ```suggestion
    $ kubectl apply --filename peering-token.yml
```

    ```

1. In “cluster-02,” create the `PeeringDialer` custom resource.

    <CodeBlockConfig filename="dialer.yml">

    ```yaml
    apiVersion: consul.hashicorp.com/v1alpha1
    kind: PeeringDialer
    metadata:
      name: cluster-01 ## The name of the peer you want to connect to
    spec:
      peer:
        secret:
          name: "peering-token"
          key: "data"
          backend: "kubernetes"
    ```

    </CodeBlockConfig>

1. Apply the `PeeringDialer` resource to the second cluster.

    ```shell-session

>     $ kubectl apply -f dialer.yml

Reviewer 3: ```suggestion
    $ kubectl apply --filename dialer.yml
```

    ```

## Deploy and export cluster services

> 1. For the service in “cluster-02” that you want to export, add the following [annotations to your service files](/docs/k8s/annotations-and-labels#consul-hashicorp-com-connect-service-upstreams).

Reviewer 2: ```suggestion
1. For the service in “cluster-02” that you want to export, add the following [annotations](/docs/k8s/annotations-and-labels#consul-hashicorp-com-connect-service-upstreams) to your service's pods. This service is referred to as "backend-service" in the following steps.
```
